torch_geometric/nn/aggr/base.py

```diff
@@ -35,35 +75,6 @@ def forward(self, x: Tensor, index: Optional[Tensor] = None,
     def reset_parameters(self):
         pass
 
-    def __call__(self, x: Tensor, index: Optional[Tensor] = None,
```

Comment:
Just so I understand better: what was the issue with this code to begin with? It looks like the tracer has problems when overwriting __call__. Is that correct? Is there any way to fix this in the tracer?

Reply:
yes, its because of this: https://github.com/pytorch/pytorch/blob/e68583b4d180066b8e4f108e0d23176a2676421c/torch/fx/_symbolic_trace.py#L702
the tracer only looks for a leave module in the patched methods - but since we overwrite call its no longer patched (the `super().__call__`  is still patched)

Reply:
Can we patch it ourselves?

Reply:
hmm, yes... but we'll have to copy quite a bit of code from their implementation. I can implement that and we can see what it looks like?

Reply:
Ok, depends on how easy it would be to try out and how large the code to patch would be. Do not burn too much time into it if it is not easy to fix :)

Reply:
I'll give it a go, and abandon if I am up too late 😁

Reply:
I pushed a working version (its a bit messy - I can cleanup what I can if we want to use it, so no need to review for style yet)

Additionally, I raised a question on the PyTorch slack channel to see if anyone had suggestions but nothing came back yet. However there is an open issue I can try and help with in pytorch to support kwargs in pre-forward-hooks (which I think would be a nice solution): https://github.com/pytorch/pytorch/issues/35643#issuecomment-1191258877

So I'd say the options now are:
- change the interface so subclasses implement `_forward`
- use this nasty approach in `fx` until we have a better solution (like the pre-forward-hooks)
- depreciate global pooling support in the hetero conversion for now and then come back to it after the above

WDYT?

Reply:
I actually think the
approach of patching isn‘t too bad, thanks for digging into this. You rock!

i was wondering if we really need to duplicate the full method here or can work with `super` instead. It looks like we only really need to do:
```python
            @st.functools.wraps(st._orig_module_call)
             def module_call_wrapper(mod, *args, **kwargs):
                 def forward(*args, **kwargs):
                     return st._orig_module_call(mod, *args, **kwargs)

                 st._autowrap_check(
                     patcher,
                     getattr(getattr(mod, "forward", mod), "__globals__", {}),
                     self._autowrap_function_ids)
                 return self.call_module(mod, forward, args, kwargs)

             with st._Patcher() as patcher:
                  patcher.patch_method(Aggregation, "__call__",
                                      module_call_wrapper, deduplicate=False)
```

Reply:
I thought about this, but I don't know how the `_Patcher` context manager works, so I was afraid to suggest it even if it happened to pass tests. 

Could certainly invest time in understanding if you think its a big upgrade thouh.

Reply:
Can we quickly test if it works? I am always in favor of reducing number of lines :)

Reply:
Hmm, no doesn't work 'out of the box'.

Reply:
Ok, then leave it as it is. Might be good to add some exhaustive documentation here, and also link to the PyTorch file this code is coming from.

Reply:
Agreed. Will keep an eye open for better solutions too.